File: app/src/main/java/autogravity/com/mvvmdemo/presentation/common/BaseActivity.java
package autogravity.com.mvvmdemo.presentation.common;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import autogravity.com.mvvmdemo.di.component.MainComponent;
import autogravity.com.mvvmdemo.presentation.MvvmApp;
import butterknife.ButterKnife;

public abstract class BaseActivity extends AppCompatActivity {

    @Override protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayoutId());
    }

    @Override public void setContentView(int layoutResID) {
        super.setContentView(layoutResID);
        ButterKnife.bind(this);
    }

    private MvvmApp getMvvmApp() {
        return (MvvmApp) getApplication();
    }

    protected MainComponent getMainComponent() {
        return getMvvmApp().getMainComponent();
    }

    protected abstract int getLayoutId();
}

File: app/src/main/java/autogravity/com/mvvmdemo/presentation/common/BaseActivity.kt
package autogravity.com.mvvmdemo.presentation.common

import android.support.v7.app.AppCompatActivity
import autogravity.com.mvvmdemo.di.component.MainComponent
import autogravity.com.mvvmdemo.presentation.MvvmApp

abstract class BaseActivity : AppCompatActivity() {

    private val mvvmApp: MvvmApp
        get() = application as MvvmApp

    protected val mainComponent: MainComponent?
        get() = mvvmApp.mainComponent
}
